rclcpp/include/rclcpp/service.hpp

```diff
@@ -486,6 +486,14 @@ class Service
   {
     rcl_ret_t ret = rcl_send_response(get_service_handle().get(), &req_id, &response);
 
+    if (ret == RCL_RET_TIMEOUT) {
+      RCLCPP_WARN(
+        rclcpp::get_node_logger(node_handle_.get()).get_child("rclcpp"),
+        "failed to send response: %s",
+        rcl_get_error_string().str);
+      rcl_reset_error();
+      return;
+    }
```

Comment:
The reason that i did not merge https://github.com/ros2/rcl/pull/1048 is how client should behave. It is easy to drop the request on the service side, but how we can manage the client still waiting on the request? I believe that needs to be considered before pushing this change into client libraries. any thoughts?

Reply:
This is another problem, and should be solved on the response subscription part.

Currently the service is suffering from a serious vulnerable, any network service should be robust enough to protect itself from malicous client attacks, or else itself can be a DoS target.

So I do consider this being a top priority issue rather than fixing the client part.

Reply:
@kghost thanks for iterating.

actually i was concerned that pending request would be the new problem for client side. but i believe that client can manage the timeout using user callback to check the future, or `spin_until_future_complete`. and then, client can remove the pending request from the list.

i would like to hear more opinions. @ros2/team 

- `rclcpp` (or `rclpy` client) should not retry the sending the response since underlying RMW implementation has done that already. (since it returns `timeout`, RMW implementation is responsible to retry.)
- if it gets `timeout`, specific exception should be thrown. Not just printing `warning` message here? this is the one of the exception, so that user application should be able to receive that? (this requires all dependent packages need to be adjusted so to avid the original issue.)

> malicous client attacks, or else itself can be a DoS target.

Even we fix this problem with this PR, this will be still a security problem? once it comes to this kind of requirement, i would use security feature, https://docs.ros.org/en/rolling/Concepts/About-Security.html ?

Reply:
We talked about this briefly in the internal meeting.  The basic problem is that we have no way to report the failure of `send_response` to the user.  However, I also agree that we should not crash the executor in this case.  So I think the best we can do at the moment is to recognize the failure, print a warning to the console, and continue on.